client/src/main/java/io/pravega/client/tables/TableSegment.java

```diff
@@ -0,0 +1,130 @@
+
+    /**
+     * Removes the given key from this Table Segment.
+     * @param key  The Key to remove. If {@link VersionedKey#getVersion()} is null, this will perform a Blind Update,
+     *             otherwise it will perform a Conditional Update based on the information provided.
+     * @return A CompletableFuture that, when completed, will indicate the Key has been removed.
+     */
+    CompletableFuture<Void> remove(VersionedKey<KeyT> key);
+
+    /**
+     * Removes one or more keys from this Table Segment.
+     *
+     * @param keys A list of keys to remove. If for at least one such key, {@link VersionedKey#getVersion()} returns a
+     *             non-null value, this will perform an atomic Conditional Update where all the keys either get removed
+     *             or none will; otherwise a Blind Update will be performed.
+     * @return A CompletableFuture that, when completed, will indicate that the keys have been removed.
+     */
+    CompletableFuture<Void> remove(List<VersionedKey<KeyT>> keys);
+
+    /**
+     * Gets the latest value for the given Key.
+     *
+     * @param key The Key to get the value for.
+     * @return A CompletableFuture that, when completed, will contain the requested result. If no such Key exists, this
+     * will be completed with a null value.
+     */
+    CompletableFuture<GetResult> get(KeyT key);
+
+    /**
+     * Gets the latest values for the given Keys.
+     *
+     * @param keys A list of Keys to get values for.
+     * @return A CompletableFuture that, when completed, will contain a list of results for the given Keys. The results
+     * will be in the same order as the keys. For any key that does not exist, a null value will be at its corresponding
+     * index.
+     */
+    CompletableFuture<List<GetResult<ValueT>>> get(List<KeyT> keys);
+
+    /**
+     * Registers a listener for all updates to a single key.
+     *
+     * @param listener The listener to register.
+     * @return A CompletableFuture that, when completed, will indicate the listener has been registered.
+     */
+    CompletableFuture<Void> registerListener(KeyUpdateListener<KeyT, ValueT> listener);
+
+    /**
+     * Unregisters a registered listener.
+     *
+     * @param listener THe listener to unregister.
+     */
+    void unregisterListener(KeyUpdateListener<KeyT, ValueT> listener);
+
+    /**
+     * Creates a new Iterator over all the Entries in the Table Segment.
+     *
+     * @param continuationToken A continuation token that can be used to resume a previously interrupted iteration. This
+     *                          can be obtained by invoking {@link IteratorItem#getContinuationToken()}.
+     * @return A CompletableFuture that, when completed, will return an {@link AsyncIterator} that can be used to iterate
+     * over all the Entries in this Table Segment.
+     */
+    CompletableFuture<AsyncIterator<IteratorItem<KeyT, ValueT>>> iterator(Object continuationToken);
```

Comment:
Does this assume ordered list of keys and guarantee that any new key added to the table will always be read once the iterator starts reading.

Reply:
No guarantees about ordering.
When you start reading, assuming you don't have any interruptions (i.e., don't use continuation token), then what you read should be a snapshot of what the index was at the time when you invoked it. 

However, with very large indices, chances are your iterator will not be able to keep up with the compaction work on the server-side so it may include newer values for the keys as they are updated. In any case, you will not get the same key twice nor will you be prevented from modifying the table while iterating.

If you want a guaranteed consistent iteration, then you need to block all of your updates to the table while this iterator is running.